pkg/services/ngalert/schedule.go

```diff
@@ -0,0 +1,215 @@
+				}
+				s, err := df[0].StringTable(-1, -1)
+				if err != nil {
+					ng.log.Info("alert definition dataframe result conversion to string failed", "definitionID", definitionID, "duration", lastEvalEnded.Sub(lastEvalStarted), "error", err)
+				}
+				ng.log.Info("alert definition result", "definitionID", definitionID, "duration", lastEvalEnded.Sub(lastEvalStarted), "res", s)
+				evalRunning = false
+			}
+		case id := <-ng.schedule.stop:
+			if id == definitionID {
+				// TODO what if it's running
+				return
+			}
+		}
+	}
+}
+
+func (ng *AlertNG) fetchAlertDefinitions(since time.Time) []*AlertDefinition {
+	cmd := listAlertDefinitionsCommand{}
+	err := ng.getAlertDefinitions(&cmd) // tmp
+	if err != nil {
+		ng.log.Error("failed to fetch updated alert definitions", "since", since, "err", err)
+		return nil
+	}
+	return cmd.Result
+}
+
+type schedule struct {
+	// base tick rate (fastest possible configured check)
+	baseInterval time.Duration
+
+	// each alert definition gets its own channel and
+	// routine. Will need lock as well. A map so when can
+	// update a specific routine if it is
+	channelMap channelMap
+
+	// broadcast channel for stopping definition routines
+	stop chan int64
+
+	cancelFunc context.CancelFunc
+
+	ctx context.Context
+
+	alertDefinitionsRunning sync.WaitGroup
+}
+
+func (sched *schedule) stopDefinitionRoutine(definitionID int64) {
+	sched.stop <- definitionID
+}
+
+// Run starts the scheduler
+func (ng *AlertNG) Run(ctx context.Context) error {
+	ng.log.Debug("ngalert starting")
+	heartbeat := time.NewTicker(ng.schedule.baseInterval)
+	var lastFetchTime time.Time
+	for {
+		select {
+		case tick := <-heartbeat.C:
+			alertDefinitions := ng.fetchAlertDefinitions(lastFetchTime)
+			ng.log.Debug("alert definitions fetched", "count", len(alertDefinitions))
+
+			// this is used for identify deleted alert definitions
+			registeredDefinitions := ng.schedule.channelMap.keyMap()
+
+			type readyToRunItem struct {
+				id           int64
+				definitionCh definitionCh
+			}
+			readyToRun := make([]readyToRunItem, 0)
+			for _, item := range alertDefinitions {
+				newRoutine := !ng.schedule.channelMap.exists(item.Id)
+				definitionCh := ng.schedule.channelMap.getOrCreateChannel(item.Id)
+
+				if newRoutine {
+					go ng.definitionRoutine(item.Id, definitionCh.ch)
+				}
+
+				if tick.Unix()%item.Interval == 0 {
+					readyToRun = append(readyToRun, readyToRunItem{id: item.Id, definitionCh: definitionCh})
+				}
+			}
+
+			if len(readyToRun) == 0 {
+				continue
+			}
+
+			step := int(ng.schedule.baseInterval.Nanoseconds()) / len(readyToRun)
+
+			// send loop is only required for distribute evaluations across time within an interval
+			for _, item := range readyToRun {
+				item.definitionCh.ch <- &evalContext{now: tick}
+
+				delete(registeredDefinitions, item.id)
+
+				time.Sleep(time.Duration(step))
```

Comment:
Using thread sleep feels like not the optimal way to distribute evaluations. 

the old scheduled seperated the scheduler tick, from actually scheduling jobs. Maybe that would help?
https://github.com/grafana/grafana/blob/master/pkg/services/alerting/scheduler.go

Reply:
I've seen some variations around on the idea of "Clock", to make sure ticker can be mocked. For example https://godoc.org/github.com/facebookgo/clock

Some discussion around something like this in the time package https://github.com/golang/go/issues/8869, blog post at https://dmitryfrank.com/articles/mocking_time_in_go , using something like this might help us with the testing here.

Reply:
Oh actually, or current alerting is using a clock https://github.com/grafana/grafana/blob/master/pkg/services/alerting/ticker.go "github.com/benbjohnson/clock"